compiler/compiler_test.go

```diff
@@ -119,28 +502,146 @@ func compile(path string, sourceFiles []source, minify bool) ([]byte, error) {
 		},
 	}
 
-	a, err := importContext.Import(path)
-	if err != nil {
-		return nil, err
-	}
-	b, err := renderPackage(a)
+	_, err := importContext.Import(root.PkgPath)
 	if err != nil {
-		return nil, err
+		t.Fatal(`failed to compile:`, err)
 	}
-	return b, nil
+	return archiveCache
 }
 
-func renderPackage(archive *Archive) ([]byte, error) {
+func renderPackage(t *testing.T, archive *Archive, minify bool) []byte {
+	t.Helper()
 	selection := make(map[*Decl]struct{})
 	for _, d := range archive.Declarations {
 		selection[d] = struct{}{}
 	}
 
 	buf := &bytes.Buffer{}
 
-	if err := WritePkgCode(archive, selection, goLinknameSet{}, false, &SourceMapFilter{Writer: buf}); err != nil {
-		return nil, err
+	if err := WritePkgCode(archive, selection, goLinknameSet{}, minify, &SourceMapFilter{Writer: buf}); err != nil {
+		t.Fatal(err)
 	}
 
-	return buf.Bytes(), nil
+	return buf.Bytes()
+}
+
+type selectionTester struct {
+	t            *testing.T
+	mainPkg      *Archive
+	archives     map[string]*Archive
+	packages     []*Archive
+	dceSelection map[*Decl]struct{}
+
+	DeclCode       *selectionCodeTester
+	InitCode       *selectionCodeTester
+	MethodListCode *selectionCodeTester
+}
+
+func declSelection(t *testing.T, sourceFiles []source, auxFiles []source) *selectionTester {
+	t.Helper()
+	root := parseSources(t, sourceFiles, auxFiles)
+	archives := compileProject(t, root, false)
+	mainPkg := archives[root.PkgPath]
+
+	paths := make([]string, 0, len(archives))
+	for path := range archives {
+		paths = append(paths, path)
+	}
+	sort.Strings(paths)
+	packages := make([]*Archive, 0, len(archives)-1)
+	for _, path := range paths {
+		packages = append(packages, archives[path])
+	}
+
+	sel := &dce.Selector[*Decl]{}
+	for _, pkg := range packages {
+		for _, d := range pkg.Declarations {
+			sel.Include(d, false)
+		}
+	}
+	dceSelection := sel.AliveDecls()
+
+	st := &selectionTester{
+		t:            t,
+		mainPkg:      mainPkg,
+		archives:     archives,
+		packages:     packages,
+		dceSelection: dceSelection,
+	}
+
+	st.DeclCode = &selectionCodeTester{st, `DeclCode`, func(d *Decl) []byte { return d.DeclCode }}
+	st.InitCode = &selectionCodeTester{st, `InitCode`, func(d *Decl) []byte { return d.InitCode }}
+	st.MethodListCode = &selectionCodeTester{st, `MethodListCode`, func(d *Decl) []byte { return d.MethodListCode }}
```

Comment:
Why assign separately instead of using the literal above?

Reply:
I just passed `st` into the `selectionCodeTester` so that I didn't have to pass in several of `selectionTester`'s fields individually, but that meant if it was in the literal I'd get a `st` not defined. It looks like all that `selectionCodeTester` uses is `t *testing.T`, `packages []*Archive`, and `dceSelection map[*Decl]struct{}` so it wouldn't actually be too much to send those in separately, so I'll just do that instead.

Reply:
Ah, makes sense, I missed the reference to `st`. Either way would have been fine then.